src/components/DiffView/index.spec.tsx

```diff
@@ -97,15 +110,54 @@ describe(__filename, () => {
         selected_file: path,
       },
     });
+
     const linterResult = createFakeExternalLinterResult({
       messages: messages.map((msg) => {
         return { ...msg, file: path };
       }),
     });
 
+    store.dispatch(
+      linterActions.loadLinterResult({
+        versionId: version.id,
+        result: linterResult,
+      }),
+    );
+
     return { linterResult, version };
   };
 
+  const renderAndGetWidgets = (params: RenderParams): WidgetMap => {
+    const root = render(params);
+
+    const diffView = root.find(Diff);
+    expect(diffView).toHaveProp('widgets');
+
+    const widgets = diffView.prop('widgets');
+    if (!widgets) {
+      throw new Error('The widgets prop was falsy');
+    }
+
+    return widgets;
+  };
+
+  const renderWidget = (hunks: Hunks, widgets: WidgetMap, line: number) => {
+    const result = getAllHunkChanges(hunks).filter(
+      (change) => change.lineNumber === line,
+    );
+    if (result.length !== 1) {
+      throw new Error(`Could not find a change at line ${line}`);
+    }
+
+    const key = getChangeKey(result[0]);
+    return shallow(<div>{widgets[key]}</div>);
+  };
+
+  const getWidgetNodes = (widgets: WidgetMap) => {
+    // Return a list of truthy React nodes in arbitrary order.
+    return Object.values(widgets).filter(Boolean);
```

Comment:
I believe `.filter((widgeet) => widget !== null)` would be slightly more readable. Could we have `widget` assigned to any other value?

Reply:
Heh. I had that at first but it felt a bit fragile since react would also not render `undefined`. I was really just trying to mimic what react would do with a list of `React.ReactNode` types.

Reply:
ah ok, it took me some time to understand what this filter was doing and the comment did not help much. That being said, now that I understand what the filter does, the comment is relevant.